### DISCUSSION TITLE
On optimizing error paths

### DISCUSSION THREAD

**Discussion by stevenroose** [rust-bitcoin/rust-bitcoin#1812] [General]
Recently I've optimization of the error path come up [a few](https://github.com/rust-bitcoin/rust-bitcoin/pull/1400#discussion_r1027355831) [times](https://github.com/rust-bitcoin/rust-bitcoin/issues/1802#issuecomment-1515000167) as an argument for certain API decisions.

The balance between devex and optimization is already often quite difficult, I would like to suggest that when it comes to error paths, optimizations should really not be that important, especially not if they potentially reduce devex.

(Judging by the amount of `Box<dyn Error + ...>` in the stdlib, I would think that avoiding certain allocations in error paths is really not an ecosystem priority..)

**Comment by apoelstra** [comment]
We mostly use `core`, not `std`, in which there are no `Box<dyn Error>` returns.

**Reply by stevenroose** [reply]
I know, I just wanted to point to stdlib's habbit of also not seeming to care too much about optimizing the error path.

**Comment by Kixunil** [comment]
I mostly agree, though there are some trivial cases that seems to be worth it to me. In particular using `AsRef<str> + Into<String>` instead of just `AsRef<str>`. It only harms exotic types and only very little.

**Reply by stevenroose** [reply]
Interestingly, you're assuming that `Into<String>` here is costless, and different from `ToString`. Which seems to align with the argument I made regarding `TryFrom<str>` vs `FromStr`, where in `TryFrom<str>` you rather expect a costless conversion while `FromStr` does parsing and might allocate other data.

Like our `Address` doesn't implement `Into<String>` but it does `ToString` (i.e. `Display`).

**Reply by Kixunil** [reply]
It's only costless for `String` -> `String`. `ToString` cannot possibly be costless because it takes `&self` and returns an owned type.

I don't think it has anything to do with `TryFrom` and I never assume `TryFrom` to be costless (nor `From`/`Into` outside specific cases actually).

And yeah, I think `Into<String>` for address would be fine. Not sure how useful. For `TryFrom` I have some specific use cases.

**Reply by stevenroose** [reply]
In the end it boils down to a semantic argument of "what is type conversion and what is serializing/parsing". I don't think practically having one of the other makes a significant difference for the user. Some, but not all, of the `std::convert` methods explicitly mention "cheap, value-to-value" conversions, which I interpret as at most shuffling around some bits and not really "doing something" like serialization.

A good example I think is how `serde_json::Value` has both `From<String>`, `From<&str>` and `FromStr`, but the semantics of the first two are significantly different (converting into `Value::String`) than the latter (parsing utf8 into a `Value`). So using the methods on a string like `{"a": true}` will result in a `Value::String('{"a": bool}')` for the former and `Value::Object(Map...)` for the latter.

So they definitely identify a semantic difference between the traits.

**Reply by apoelstra** [reply]
> A good example I think is how `serde_json::Value` has both `From<String>`, `From<&str>` and `FromStr`, but the semantics of the first two are significantly different (converting into `Value::String`) than the latter (parsing utf8 into a `Value`). So

This is a really interesting example.

**Reply by Kixunil** [reply]
> In the end it boils down to a semantic argument of "what is type conversion and what is serializing/parsing". I don't think practically having one of the other makes a significant difference for the user. Some, but not all, of the `std::convert` methods explicitly mention "cheap, value-to-value" conversions, which I interpret as at most shuffling around some bits and not really "doing something" like serialization.
> 
> A good example I think is how `serde_json::Value` has both `From<String>`, `From<&str>` and `FromStr`, but the semantics of the first two are significantly different (converting into `Value::String`) than the latter (parsing utf8 into a `Value`). So using the methods on a string like `{"a": true}` will result in a `Value::String('{"a": bool}')` for the former and `Value::Object(Map...)` for the latter.
> 
> So they definitely identify a semantic difference between the traits.

Yes, only reference conversions are guaranteed to be cheap. These are unrelated to current discussion.



> In the end it boils down to a semantic argument of "what is type conversion and what is serializing/parsing". I don't think practically having one of the other makes a significant difference for the user. Some, but not all, of the `std::convert` methods explicitly mention "cheap, value-to-value" conversions, which I interpret as at most shuffling around some bits and not really "doing something" like serialization.
> 
> A good example I think is how `serde_json::Value` has both `From<String>`, `From<&str>` and `FromStr`, but the semantics of the first two are significantly different (converting into `Value::String`) than the latter (parsing utf8 into a `Value`). So using the methods on a string like `{"a": true}` will result in a `Value::String('{"a": bool}')` for the former and `Value::Object(Map...)` for the latter.
> 
> So they definitely identify a semantic difference between the traits.

I believe this to be a foot gun but it's not going to change. Anyway even if there is a semantic difference I don't see any harm in implementing both and making them do the same thing in our library.

**Reply by stevenroose** [reply]
Like I think someone else mentioned somewhere, I think a lot of the confusion would be gone if `FromStr` was just named `Parse`. Specifically in `Value`, the only confusion I can see is `Value::from_str` doing JSON parsing and not converting into `Value::String`. But if you use `"{}".parse()` it is more intuitive, which is what the `FromStr` was kinda made for.

**Reply by Kixunil** [reply]
Yes. `s.parse()` is idiomatic. Problems tend to appear when not using idiomatic code. :)
